actionpack/lib/action_dispatch/routing/route_set.rb

```diff
@@ -144,6 +159,62 @@ def length
           routes.length
         end
 
+        def add_url_helper(name, defaults, &block)
+          @custom_helpers << name
+          helper = CustomUrlHelper.new(name, defaults, &block)
+
+          @path_helpers_module.module_eval do
+            define_method(:"#{name}_path") do |*args|
+              options = args.extract_options!
+              helper.call(self, args, options, only_path: true)
+            end
+          end
+
+          @url_helpers_module.module_eval do
+            define_method(:"#{name}_url") do |*args|
+              options = args.extract_options!
+              helper.call(self, args, options)
+            end
+          end
+        end
+
+        class CustomUrlHelper
+          attr_reader :name, :defaults, :block
+
+          def initialize(name, defaults, &block)
+            @name = name
+            @defaults = defaults
+            @block = block
+          end
+
+          def call(t, args, options, outer_options = {})
+            url_options = eval_block(t, args, options)
+
+            case url_options
```

Comment:
It's similar to the behaviour in ActionDispatch::Routing::UrlFor but not exactly the same, mainly because of the options being needed to be combined into the arguments for `url_for`.